<a href="https://colab.research.google.com/github/AliRaddman/divar-ml-project/blob/main/notebooks/work/prep_ben.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



*   imports
*   mount drive

*   load cleaned_step1.parquet






In [ ]:
from google.colab import drive
drive.mount('/content/drive')

import pandas as pd
import numpy as np
import os

pd.set_option("display.max_columns", 150)
pd.set_option("display.max_rows", 100)
pd.set_option("display.width", 250)

Mounted at /content/drive


In [ ]:
CLEAN_PATH = "/content/drive/MyDrive/divar_project/cleaned_step1.parquet"

print("clean path:", CLEAN_PATH)
print("exists:", os.path.exists(CLEAN_PATH))

if os.path.exists(CLEAN_PATH):
    print("size MB:", round(os.path.getsize(CLEAN_PATH) / 1024**2, 2))
else:
    print("File not found. Check Drive shortcut or path.")

clean path: /content/drive/MyDrive/divar_project/cleaned_step1.parquet
exists: True
size MB: 291.92


 - shape
 - schema
 - describe numeric columns
- boolean columns
- geo check

In [ ]:
df = pd.read_parquet(CLEAN_PATH)

print("Loaded cleaned parquet")
print("shape:", df.shape)
print("memory MB:", round(df.memory_usage(deep=True).sum() / 1024**2, 2))

display(df.head(5))

Loaded cleaned parquet
shape: (999943, 53)
memory MB: 1028.58


,cat2_slug,cat3_slug,city_slug,neighborhood_slug,created_at_month,user_type,description,title,rent_mode,rent_value,rent_type,price_mode,price_value,credit_mode,credit_value,rent_credit_transform,transformable_price,transformable_credit,transformed_credit,transformable_rent,transformed_rent,land_size,building_size,deed_type,has_business_deed,floor,rooms_count,total_floors_count,unit_per_floor,has_balcony,has_elevator,has_warehouse,has_parking,construction_year,is_rebuilt,has_warm_water_provider,has_heating_system,has_cooling_system,has_restroom,has_security_guard,has_barbecue,building_direction,has_pool,has_jacuzzi,has_sauna,floor_material,property_type,regular_person_capacity,location_latitude,location_longitude,location_radius,created_at_shamsi,created_at_shamsi_readable
0,temporary-rent,villa,karaj,mehrshahr,2024-08-01,مشاور املاک,۵۰۰متر\n۲۰۰متر بنا دوبلکس\n۳خواب\nاستخر آبگرم ...,باغ ویلا اجاره روزانه استخر داخل لشکرآباد سهیلیه,NaN,NaN,NaN,NaN,NaN,NaN,NaN,<NA>,<NA>,NaN,NaN,NaN,NaN,NaN,500.0,NaN,<NA>,NaN,3.0,NaN,NaN,False,False,False,False,1395.0,False,NaN,NaN,NaN,NaN,False,False,NaN,False,False,False,NaN,NaN,4.0,35.811684,50.936600,500.0,1403-05,مرداد 1403
1,residential-sell,apartment-sell,tehran,gholhak,2024-05-01,مشاور املاک,دسترسی عالی به مترو و شریعتی \nمشاعات تمیز \nب...,۶۰ متر قلهک فول امکانات,NaN,NaN,NaN,مقطوع,8.500000e+09,NaN,NaN,<NA>,<NA>,NaN,NaN,NaN,NaN,NaN,60.0,NaN,<NA>,3.0,1.0,NaN,NaN,False,True,True,True,1384.0,False,NaN,NaN,NaN,NaN,False,False,NaN,False,False,False,NaN,NaN,NaN,NaN,NaN,500.0,1403-02,اردیبهشت 1403
2,residential-rent,apartment-rent,tehran,tohid,2024-10-01,NaN,تخلیه پایان ماه,آپارتمان ۳ خوابه ۱۳۲ متر,مقطوع,26000000.0,NaN,NaN,NaN,مقطوع,750000000.0,False,False,750000000.0,NaN,26000000.0,NaN,NaN,132.0,NaN,<NA>,3.0,3.0,NaN,NaN,False,True,True,True,1401.0,False,NaN,NaN,NaN,NaN,False,False,NaN,False,False,False,NaN,NaN,NaN,35.703865,51.373459,NaN,1403-07,مهر 1403
3,commercial-rent,office-rent,tehran,elahiyeh,2024-06-01,NaN,فرشته تاپ لوکیشن\n۹۰ متر موقعیت اداری\nیک اتاق...,فرشته ۹۰ متر دفتر کار مدرن موقعیت اداری,مقطوع,95000000.0,NaN,NaN,NaN,مقطوع,950000000.0,False,False,950000000.0,NaN,95000000.0,NaN,NaN,90.0,NaN,<NA>,4.0,1.0,NaN,NaN,False,True,False,True,1400.0,False,NaN,NaN,NaN,NaN,False,False,NaN,False,False,False,NaN,NaN,NaN,NaN,NaN,NaN,1403-03,خرداد 1403
4,residential-sell,apartment-sell,mashhad,emamreza,2024-05-01,مشاور املاک,هلدینگ ساختمانی اکبری\n\nهمراه شما هستیم برای ...,۱۱۵ متری/شمالی رو به آفتاب/اکبری,NaN,NaN,NaN,مقطوع,5.750000e+09,NaN,NaN,<NA>,<NA>,NaN,NaN,NaN,NaN,NaN,115.0,single_page,<NA>,4.0,2.0,6.0,NaN,True,True,True,True,1403.0,False,package,shoofaj,air_conditioner,squat_seat,False,False,north,False,False,False,ceramic,NaN,NaN,NaN,NaN,NaN,1403-02,اردیبهشت 1403


In [ ]:
schema = pd.DataFrame({
    "column": df.columns,
    "dtype": [str(df[col].dtype) for col in df.columns],
    "non_null_count": [df[col].notna().sum() for col in df.columns],
    "null_count": [df[col].isna().sum() for col in df.columns],
    "null_percent": [round(df[col].isna().mean() * 100, 2) for col in df.columns],
    "unique_count": [df[col].nunique(dropna=True) for col in df.columns],
})

display(schema)

,column,dtype,non_null_count,null_count,null_percent,unique_count
0,cat2_slug,category,999943,0,0.00,6
1,cat3_slug,category,999943,0,0.00,16
2,city_slug,category,999943,0,0.00,421
3,neighborhood_slug,category,437089,562854,56.29,1188
4,created_at_month,datetime64[ns],999943,0,0.00,46
5,user_type,category,288877,711066,71.11,2
6,description,object,999943,0,0.00,989397
7,title,object,999943,0,0.00,904089
8,rent_mode,category,352961,646982,64.70,3
9,rent_value,float64,351290,648653,64.87,1311


In [ ]:
important_numeric_cols = [
    "price_value",
    "rent_value",
    "credit_value",
    "transformable_price",
    "transformable_credit",
    "transformed_credit",
    "transformable_rent",
    "transformed_rent",
    "land_size",
    "building_size",
    "floor",
    "rooms_count",
    "total_floors_count",
    "unit_per_floor",
    "construction_year",
    "location_latitude",
    "location_longitude",
    "location_radius"
]

existing_numeric_cols = [col for col in important_numeric_cols if col in df.columns]

print("Existing important numeric columns:")
print(existing_numeric_cols)

display(
    df[existing_numeric_cols]
    .describe(percentiles=[0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99])
    .T
)

Existing important numeric columns:
['price_value', 'rent_value', 'credit_value', 'transformable_price', 'transformable_credit', 'transformed_credit', 'transformable_rent', 'transformed_rent', 'land_size', 'building_size', 'floor', 'rooms_count', 'total_floors_count', 'unit_per_floor', 'construction_year', 'location_latitude', 'location_longitude', 'location_radius']


,count,mean,std,min,1%,5%,25%,50%,75%,95%,99%,max
price_value,568330.0,1.736564e+10,5.878822e+11,0.000000,3.400000e+03,2.600000e+08,1.400000e+09,2.840000e+09,5.900000e+09,2.100000e+10,8.500000e+10,1.000000e+14
rent_value,351290.0,4.102673e+10,3.807708e+12,0.000000,0.000000e+00,0.000000e+00,1.111110e+05,5.000000e+06,1.200000e+07,5.500000e+07,2.200000e+08,1.000000e+15
credit_value,352063.0,4.872522e+10,4.341543e+12,0.000000,1.000000e+06,1.500000e+07,1.000000e+08,2.500000e+08,5.000000e+08,1.650000e+09,5.000000e+09,1.000000e+15
transformable_credit,352052.0,4.872674e+10,4.341611e+12,0.000000,1.000000e+06,1.500000e+07,1.000000e+08,2.500000e+08,5.000000e+08,1.650000e+09,5.000000e+09,1.000000e+15
transformed_credit,72406.0,8.557358e+09,2.064619e+12,0.000000,5.000000e+07,1.000000e+08,2.000000e+08,4.000000e+08,8.500000e+08,2.900000e+09,8.000000e+09,5.555556e+14
transformable_rent,351215.0,4.103549e+10,3.808114e+12,0.000000,0.000000e+00,0.000000e+00,1.111110e+05,5.000000e+06,1.200000e+07,5.500000e+07,2.200000e+08,1.000000e+15
transformed_rent,72406.0,1.619943e+07,5.217990e+07,0.000000,0.000000e+00,0.000000e+00,1.000000e+06,6.000000e+06,1.500000e+07,5.800000e+07,1.800000e+08,3.000000e+09
land_size,186390.0,4.165602e+03,1.218947e+05,1.000000,4.000000e+01,6.000000e+01,1.100000e+02,1.950000e+02,2.800000e+02,1.000000e+03,4.321430e+03,1.000000e+07
building_size,999943.0,4.355842e+03,1.353702e+05,1.000000,1.200000e+01,3.800000e+01,7.500000e+01,1.030000e+02,1.620000e+02,8.000000e+02,1.000000e+04,1.000000e+07
floor,541031.0,2.702769e+00,2.208308e+00,-1.000000,0.000000e+00,0.000000e+00,1.000000e+00,2.000000e+00,4.000000e+00,6.000000e+00,1.100000e+01,3.000000e+01


In [ ]:
cols_to_check = [
    "price_value",
    "rent_value",
    "credit_value",
    "transformable_credit",
    "transformed_credit",
    "transformable_rent",
    "transformed_rent",
    "land_size",
    "building_size",
    "construction_year"
]

cols_to_check = [col for col in cols_to_check if col in df.columns]

for col in cols_to_check:
    print("=" * 100)
    print("COLUMN:", col)
    print("dtype:", df[col].dtype)
    print("missing:", df[col].isna().sum(), "|", round(df[col].isna().mean() * 100, 2), "%")

    if pd.api.types.is_numeric_dtype(df[col]):
        print("zero count:", (df[col] == 0).sum())
        print("negative count:", (df[col] < 0).sum())
        print("min:", df[col].min())
        print("max:", df[col].max())

        print("\nSmallest values:")
        display(df[[col]].sort_values(col, ascending=True).head(10))

        print("\nLargest values:")
        display(df[[col]].sort_values(col, ascending=False).head(10))
    else:
        print("Not numeric")

COLUMN: price_value
dtype: float64
missing: 431613 | 43.16 %
zero count: 1902
negative count: 0
min: 0.0
max: 99999999999999.0

Smallest values:


,price_value
411595,0.0
235798,0.0
642482,0.0
778784,0.0
848446,0.0
714873,0.0
957764,0.0
325643,0.0
528642,0.0
360099,0.0



Largest values:


,price_value
931721,1.000000e+14
786936,1.000000e+14
927237,1.000000e+14
740448,1.000000e+14
588158,1.000000e+14
329736,1.000000e+14
116850,1.000000e+14
6406,9.000000e+13
560215,8.900000e+13
338993,8.397000e+13


COLUMN: rent_value
dtype: float64
missing: 648653 | 64.87 %
zero count: 59238
negative count: 0
min: 0.0
max: 999999999999999.0

Smallest values:


,rent_value
550173,0.0
340810,0.0
550244,0.0
550235,0.0
851033,0.0
112217,0.0
112216,0.0
112191,0.0
112183,0.0
851093,0.0



Largest values:


,rent_value
235482,1.000000e+15
463445,1.000000e+15
340833,1.000000e+15
149250,8.888889e+14
906011,3.333333e+14
554350,3.333333e+14
538238,2.222222e+14
577215,1.444146e+14
881404,1.200000e+14
595801,1.111111e+14


COLUMN: credit_value
dtype: float64
missing: 647880 | 64.79 %
zero count: 2864
negative count: 0
min: 0.0
max: 999999999999999.0

Smallest values:


,credit_value
61363,0.0
212244,0.0
636714,0.0
636655,0.0
327123,0.0
169524,0.0
636311,0.0
169627,0.0
870116,0.0
870209,0.0



Largest values:


,credit_value
340833,1.000000e+15
463445,1.000000e+15
235482,1.000000e+15
370772,1.000000e+15
547693,8.888889e+14
200111,5.000000e+14
976804,4.000004e+14
217471,3.333333e+14
148554,3.000000e+14
149250,3.000000e+14


COLUMN: transformable_credit
dtype: float64
missing: 647891 | 64.79 %
zero count: 2863
negative count: 0
min: 0.0
max: 999999999999999.0

Smallest values:


,transformable_credit
943741,0.0
482720,0.0
887291,0.0
871294,0.0
857133,0.0
447118,0.0
29476,0.0
196505,0.0
816046,0.0
531775,0.0



Largest values:


,transformable_credit
463445,1.000000e+15
370772,1.000000e+15
340833,1.000000e+15
235482,1.000000e+15
547693,8.888889e+14
200111,5.000000e+14
976804,4.000004e+14
217471,3.333333e+14
149250,3.000000e+14
148554,3.000000e+14


COLUMN: transformed_credit
dtype: float64
missing: 927537 | 92.76 %
zero count: 87
negative count: 0
min: 0.0
max: 555555555555555.0

Smallest values:


,transformed_credit
254596,0.0
577330,0.0
100830,0.0
884791,0.0
844759,0.0
434861,0.0
563494,0.0
33901,0.0
562990,0.0
713646,0.0



Largest values:


,transformed_credit
255774,5.555556e+14
895180,3.500000e+11
265865,1.400000e+11
677775,1.300000e+11
773734,1.300000e+11
296733,1.250000e+11
638106,1.200000e+11
728522,1.200000e+11
358646,1.000000e+11
587691,9.700000e+10


COLUMN: transformable_rent
dtype: float64
missing: 648728 | 64.88 %
zero count: 59171
negative count: 0
min: 0.0
max: 999999999999999.0

Smallest values:


,transformable_rent
831200,0.0
421341,0.0
421372,0.0
831122,0.0
831135,0.0
831138,0.0
831140,0.0
156601,0.0
831067,0.0
545174,0.0



Largest values:


,transformable_rent
340833,1.000000e+15
463445,1.000000e+15
235482,1.000000e+15
149250,8.888889e+14
906011,3.333333e+14
554350,3.333333e+14
538238,2.222222e+14
577215,1.444146e+14
881404,1.200000e+14
474275,1.111111e+14


COLUMN: transformed_rent
dtype: float64
missing: 927537 | 92.76 %
zero count: 13074
negative count: 0
min: 0.0
max: 3000000000.0

Smallest values:


,transformed_rent
999473,0.0
999077,0.0
999129,0.0
999135,0.0
921,0.0
957,0.0
1039,0.0
998574,0.0
998654,0.0
1102,0.0



Largest values:


,transformed_rent
201931,3.000000e+09
857685,3.000000e+09
881543,2.000000e+09
295608,1.800000e+09
204259,1.800000e+09
887495,1.650000e+09
197763,1.650000e+09
723178,1.500000e+09
844759,1.500000e+09
615232,1.500000e+09


COLUMN: land_size
dtype: float64
missing: 813553 | 81.36 %
zero count: 0
negative count: 0
min: 1.0
max: 10000000.0

Smallest values:


,land_size
434551,1.0
555271,1.0
351966,1.0
345016,1.0
112083,1.0
105515,1.0
969515,1.0
957693,1.0
81379,1.0
303812,1.0



Largest values:


,land_size
173265,10000000.0
905048,10000000.0
901144,10000000.0
176030,10000000.0
734362,10000000.0
145522,10000000.0
336711,10000000.0
534216,10000000.0
449409,10000000.0
535300,10000000.0


COLUMN: building_size
dtype: float64
missing: 0 | 0.0 %
zero count: 0
negative count: 0
min: 1.0
max: 10000000.0

Smallest values:


,building_size
762122,1.0
929030,1.0
526390,1.0
843039,1.0
931097,1.0
473977,1.0
902253,1.0
993871,1.0
671427,1.0
810400,1.0



Largest values:


,building_size
170373,10000000.0
477673,10000000.0
3207,10000000.0
480281,10000000.0
999792,10000000.0
543124,10000000.0
329736,10000000.0
195796,10000000.0
903667,10000000.0
317461,10000000.0


COLUMN: construction_year
dtype: float64
missing: 0 | 0.0 %
zero count: 0
negative count: 0
min: 1365.0
max: 1403.0

Smallest values:


,construction_year
522020,1365.0
522163,1365.0
830135,1365.0
830159,1365.0
209254,1365.0
522106,1365.0
522145,1365.0
304459,1365.0
731468,1365.0
117680,1365.0



Largest values:


,construction_year
999938,1403.0
999939,1403.0
640460,1403.0
813521,1403.0
813520,1403.0
331491,1403.0
39,1403.0
42,1403.0
477959,1403.0
477966,1403.0


In [ ]:
important_categorical_cols = [
    "cat2_slug",
    "cat3_slug",
    "city_slug",
    "neighborhood_slug",
    "user_type",
    "rent_mode",
    "rent_type",
    "price_mode",
    "credit_mode",
    "deed_type",
    "property_type",
    "building_direction",
    "floor_material"
]

existing_categorical_cols = [col for col in important_categorical_cols if col in df.columns]

for col in existing_categorical_cols:
    print("=" * 100)
    print("COLUMN:", col)
    print("dtype:", df[col].dtype)
    print("unique:", df[col].nunique(dropna=True))
    print("missing:", df[col].isna().sum(), "|", round(df[col].isna().mean() * 100, 2), "%")
    display(df[col].value_counts(dropna=False).head(25))

COLUMN: cat2_slug
dtype: category
unique: 6
missing: 0 | 0.0 %


,count
cat2_slug,
residential-sell,558692
residential-rent,276528
commercial-rent,76564
commercial-sell,38860
temporary-rent,29897
real-estate-services,19402


COLUMN: cat3_slug
dtype: category
unique: 16
missing: 0 | 0.0 %


,count
cat3_slug,
apartment-sell,303372
apartment-rent,211853
plot-old,133570
house-villa-sell,121750
house-villa-rent,64675
shop-rent,45993
shop-sell,21855
office-rent,21416
suite-apartment,16460


COLUMN: city_slug
dtype: category
unique: 421
missing: 0 | 0.0 %


,count
city_slug,
tehran,190898
mashhad,68993
karaj,49367
shiraz,37139
isfahan,36951
tabriz,27654
andisheh-new-town,20847
ahvaz,19589
qom,17109


COLUMN: neighborhood_slug
dtype: category
unique: 1188
missing: 562854 | 56.29 %


,count
neighborhood_slug,
NaN,562854
elahiyehblvd,7076
poonak,5393
ghasemabad,5216
saadat-abad,4686
bolvartoos,4684
gohardasht,3738
west-tehranpars,3575
chitgar-lake,3560


COLUMN: user_type
dtype: category
unique: 2
missing: 711066 | 71.11 %


,count
user_type,
NaN,711066
مشاور املاک,256006
شخصی,32871


COLUMN: rent_mode
dtype: category
unique: 3
missing: 646982 | 64.7 %


,count
rent_mode,
NaN,646982
مقطوع,292052
مجانی,59238
توافقی,1671


COLUMN: rent_type
dtype: category
unique: 2
missing: 895985 | 89.6 %


,count
rent_type,
NaN,895985
rent_credit,103811
full_credit,147


COLUMN: price_mode
dtype: category
unique: 3
missing: 426353 | 42.64 %


,count
price_mode,
مقطوع,566428
NaN,426353
توافقی,5260
مجانی,1902


COLUMN: credit_mode
dtype: category
unique: 3
missing: 646982 | 64.7 %


,count
credit_mode,
NaN,646982
مقطوع,349199
مجانی,2864
توافقی,898


COLUMN: deed_type
dtype: category
unique: 4
missing: 747056 | 74.71 %


,count
deed_type,
NaN,747056
single_page,194261
written_agreement,31873
other,20288
multi_page,6465


COLUMN: property_type
dtype: category
unique: 5
missing: 972886 | 97.29 %


,count
property_type,
NaN,972886
jungle,16614
beach,3990
other,3832
jungle-mountain,2399
mountain,222


COLUMN: building_direction
dtype: category
unique: 4
missing: 679228 | 67.93 %


,count
building_direction,
NaN,679228
south,159453
north,142008
east,12111
west,7143


COLUMN: floor_material
dtype: category
unique: 7
missing: 595630 | 59.57 %


,count
floor_material,
NaN,595630
ceramic,307819
stone,38395
carpet,17696
mosaic,14609
wood_parquet,14419
laminate_parquet,9582
floor_covering,1793


In [ ]:
bool_like_cols = [
    col for col in df.columns
    if col.startswith("has_") or col.endswith("_has") or col.startswith("is_")
]

print("Boolean-like columns count:", len(bool_like_cols))
print(bool_like_cols)

for col in bool_like_cols:
    print("=" * 100)
    print("COLUMN:", col)
    print("dtype:", df[col].dtype)
    display(df[col].value_counts(dropna=False).head(10))

Boolean-like columns count: 15
['has_business_deed', 'has_balcony', 'has_elevator', 'has_warehouse', 'has_parking', 'is_rebuilt', 'has_warm_water_provider', 'has_heating_system', 'has_cooling_system', 'has_restroom', 'has_security_guard', 'has_barbecue', 'has_pool', 'has_jacuzzi', 'has_sauna']
COLUMN: has_business_deed
dtype: boolean


,count
has_business_deed,
<NA>,965265
False,17427
True,17251


COLUMN: has_balcony
dtype: boolean


,count
has_balcony,
False,587160
True,412783


COLUMN: has_elevator
dtype: boolean


,count
has_elevator,
False,634830
True,365113


COLUMN: has_warehouse
dtype: boolean


,count
has_warehouse,
True,582269
False,417674


COLUMN: has_parking
dtype: boolean


,count
has_parking,
True,552495
False,447448


COLUMN: is_rebuilt
dtype: boolean


,count
is_rebuilt,
False,823665
True,176278


COLUMN: has_warm_water_provider
dtype: category


,count
has_warm_water_provider,
NaN,621827
package,205641
water_heater,141080
powerhouse,31395


COLUMN: has_heating_system
dtype: category


,count
has_heating_system,
NaN,634095
shoofaj,201355
heater,119770
duct_split,22047
floor_heating,7716
split,5845
fan_coil,5065
fireplace,4050


COLUMN: has_cooling_system
dtype: category


,count
has_cooling_system,
NaN,655244
water_cooler,219264
air_conditioner,52341
split,38707
duct_split,29373
fan_coil,5014


COLUMN: has_restroom
dtype: category


,count
has_restroom,
NaN,593662
squat_seat,219392
squat,179434
seat,7455


COLUMN: has_security_guard
dtype: boolean


,count
has_security_guard,
False,993262
True,6681


COLUMN: has_barbecue
dtype: boolean


,count
has_barbecue,
False,992122
True,7821


COLUMN: has_pool
dtype: boolean


,count
has_pool,
False,996222
True,3721


COLUMN: has_jacuzzi
dtype: boolean


,count
has_jacuzzi,
False,997670
True,2273


COLUMN: has_sauna
dtype: boolean


,count
has_sauna,
False,998548
True,1395


In [ ]:
geo_cols = [
    col for col in [
        "city_slug",
        "neighborhood_slug",
        "location_latitude",
        "location_longitude",
        "location_radius"
    ]
    if col in df.columns
]

display(df[geo_cols].head(10))

if "location_latitude" in df.columns and "location_longitude" in df.columns:
    geo_df = df[["location_latitude", "location_longitude"]].dropna()

    print("geo rows:", len(geo_df))
    print("geo missing rows:", df.shape[0] - len(geo_df))
    print("latitude range:", geo_df["location_latitude"].min(), "to", geo_df["location_latitude"].max())
    print("longitude range:", geo_df["location_longitude"].min(), "to", geo_df["location_longitude"].max())

    invalid_geo = df[
        (df["location_latitude"].notna()) &
        (df["location_longitude"].notna()) &
        ~(
            df["location_latitude"].between(24, 40) &
            df["location_longitude"].between(43, 64)
        )
    ]

    print("invalid geo rows outside rough Iran bounds:", len(invalid_geo))
    display(invalid_geo[geo_cols].head(20))

,city_slug,neighborhood_slug,location_latitude,location_longitude,location_radius
0,karaj,mehrshahr,35.811684,50.936600,500.0
1,tehran,gholhak,NaN,NaN,500.0
2,tehran,tohid,35.703865,51.373459,NaN
3,tehran,elahiyeh,NaN,NaN,NaN
4,mashhad,emamreza,NaN,NaN,NaN
5,ahvaz,mellirah,NaN,NaN,NaN
6,kermanshah,NaN,NaN,NaN,NaN
7,tehran,dardasht,35.729832,51.505466,500.0
8,mahdasht-city,NaN,35.712364,50.794781,500.0
9,mashhad,faramarzabbasi,NaN,NaN,NaN


geo rows: 655594
geo missing rows: 344349
latitude range: 23.62647819519043 to 40.358055114746094
longitude range: 40.16236877441406 to 74.5116195678711
invalid geo rows outside rough Iran bounds: 18


,city_slug,neighborhood_slug,location_latitude,location_longitude,location_radius
10895,nahavand,NaN,23.636976,52.240677,NaN
71647,qeshm,NaN,23.636976,55.139008,NaN
145957,mashhad,torbatheydarieh,32.822731,74.511620,NaN
420979,fasa-city,NaN,23.636976,52.334061,NaN
460012,konarak,NaN,23.636976,59.933167,NaN
494275,bandar-abbas,NaN,23.636976,55.105362,NaN
510854,tabriz,NaN,36.005264,40.162369,NaN
565173,qeshm,NaN,23.636976,54.584198,NaN
565402,hashtgerd-city,NaN,36.983528,40.162369,NaN
672741,bandar-abbas,NaN,23.777609,55.415295,NaN


In [ ]:
needed_cols = [
    col for col in [
        "cat3_slug",
        "price_value",
        "building_size",
        "city_slug",
        "property_type"
    ]
    if col in df.columns
]

sale_df = df[needed_cols].copy()

if "price_value" in sale_df.columns:
    sale_df = sale_df[sale_df["price_value"].notna()]
    sale_df = sale_df[sale_df["price_value"] > 0]

print("sale_df shape:", sale_df.shape)

if "cat3_slug" in sale_df.columns:
    print("cat3 distribution in sale_df:")
    display(sale_df["cat3_slug"].value_counts(dropna=False).head(30))

    price_by_cat3 = sale_df.groupby("cat3_slug")["price_value"].agg(
        count="count",
        mean="mean",
        median="median",
        min="min",
        max="max"
    ).sort_values("count", ascending=False)

    display(price_by_cat3.head(30))

sale_df shape: (566428, 5)
cat3 distribution in sale_df:


,count
cat3_slug,
apartment-sell,302629
house-villa-sell,120780
plot-old,109836
shop-sell,18445
industry-agriculture-business-sell,9029
office-sell,4455
presell,1203
shop-rent,23
partnership,6


/tmp/ipykernel_2013/1391722094.py:24: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  price_by_cat3 = sale_df.groupby("cat3_slug")["price_value"].agg(


,count,mean,median,min,max
cat3_slug,,,,,
apartment-sell,302629,1.553427e+10,3.300000e+09,1.000000e+00,1.000000e+14
house-villa-sell,120780,1.551057e+10,2.700000e+09,1.000000e+00,1.000000e+14
plot-old,109836,2.243806e+10,1.350000e+09,5.000000e+04,1.000000e+14
shop-sell,18445,2.930001e+10,3.200000e+09,1.000000e+00,1.000000e+14
industry-agriculture-business-sell,9029,1.865589e+10,2.500000e+09,1.000000e+00,1.500000e+13
office-sell,4455,2.544401e+10,6.800000e+09,1.000000e+00,8.500000e+12
presell,1203,6.723954e+09,4.000000e+09,1.000000e+00,8.316000e+11
shop-rent,23,2.658700e+08,1.350000e+08,1.111100e+04,2.000000e+09
partnership,6,1.340185e+10,1.055556e+10,3.800000e+09,3.000000e+10


In [ ]:
hyp_cols = [
    col for col in [
        "construction_year",
        "building_size",
        "land_size",
        "cat3_slug",
        "property_type",
        "city_slug"
    ]
    if col in df.columns
]

hyp_df = df[hyp_cols].copy()

display(hyp_df.head(10))

display(
    hyp_df.describe(percentiles=[0.01, 0.05, 0.5, 0.95, 0.99]).T
)

if "construction_year" in hyp_df.columns:
    print("construction_year value counts:")
    display(hyp_df["construction_year"].value_counts(dropna=False).sort_index().tail(80))

if "construction_year" in hyp_df.columns and "building_size" in hyp_df.columns:
    old_new_df = hyp_df[
        hyp_df["construction_year"].notna() &
        hyp_df["building_size"].notna() &
        (hyp_df["building_size"] > 0)
    ].copy()

    old_new_df["age_group"] = np.where(
        old_new_df["construction_year"] < 1396,
        "old_before_1396",
        "new_1396_or_after"
    )

    display(
        old_new_df.groupby("age_group")["building_size"].agg(
            count="count",
            mean="mean",
            median="median",
            std="std",
            min="min",
            max="max"
        )
    )

,construction_year,building_size,land_size,cat3_slug,property_type,city_slug
0,1395.0,500.0,NaN,villa,NaN,karaj
1,1384.0,60.0,NaN,apartment-sell,NaN,tehran
2,1401.0,132.0,NaN,apartment-rent,NaN,tehran
3,1400.0,90.0,NaN,office-rent,NaN,tehran
4,1403.0,115.0,NaN,apartment-sell,NaN,mashhad
5,1389.0,100.0,NaN,apartment-rent,NaN,ahvaz
6,1395.0,80.0,NaN,office-rent,NaN,kermanshah
7,1393.0,100.0,NaN,apartment-sell,NaN,tehran
8,1396.0,78.0,NaN,apartment-sell,NaN,mahdasht-city
9,1387.0,80.0,NaN,apartment-sell,NaN,mashhad


,count,mean,std,min,1%,5%,50%,95%,99%,max
construction_year,999943.0,1393.919751,7.845562,1365.0,1365.0,1380.0,1395.0,1403.0,1403.00,1403.0
building_size,999943.0,4355.841994,135370.213004,1.0,12.0,38.0,103.0,800.0,10000.00,10000000.0
land_size,186390.0,4165.601529,121894.651525,1.0,40.0,60.0,195.0,1000.0,4321.43,10000000.0


construction_year value counts:


,count
construction_year,
1365.0,20636
1371.0,5529
1372.0,1914
1373.0,1827
1374.0,1289
1375.0,7247
1376.0,1593
1377.0,2117
1378.0,3025


,count,mean,median,std,min,max
age_group,,,,,,
new_1396_or_after,403273,3278.519556,105.0,120232.414815,1.0,10000000.0
old_before_1396,596670,5083.974884,103.0,144702.925597,1.0,10000000.0


In [ ]:
required_for_benyamin = {
    "unit_price_and_outliers": [
        "price_value",
        "rent_value",
        "credit_value",
        "building_size",
        "land_size"
    ],
    "descriptive_q4": [
        "cat3_slug",
        "price_value"
    ],
    "descriptive_q5": [
        "location_latitude",
        "location_longitude",
        "city_slug",
        "neighborhood_slug"
    ],
    "hypothesis_2": [
        "construction_year",
        "building_size"
    ]
}

for section, cols in required_for_benyamin.items():
    print("=" * 100)
    print(section)
    for col in cols:
        print(f"{col:25s}", "FOUND" if col in df.columns else "MISSING")

unit_price_and_outliers
price_value               FOUND
rent_value                FOUND
credit_value              FOUND
building_size             FOUND
land_size                 FOUND
descriptive_q4
cat3_slug                 FOUND
price_value               FOUND
descriptive_q5
location_latitude         FOUND
location_longitude        FOUND
city_slug                 FOUND
neighborhood_slug         FOUND
hypothesis_2
construction_year         FOUND
building_size             FOUND


- transaction_type
- positive_or_nan
- area_for_unit_price
- monthly_rent_equivalent
- unit prices
- outlier flags
- final report

In [ ]:

df_pre = df.copy(deep=False)

print("Original shape:", df.shape)
print("Preprocessing dataframe shape:", df_pre.shape)

Original shape: (999943, 53)
Preprocessing dataframe shape: (999943, 53)


In [ ]:
cat2 = df_pre["cat2_slug"].astype("string")
cat3 = df_pre["cat3_slug"].astype("string")

df_pre["transaction_type"] = np.select(
    [
        cat2.eq("temporary-rent"),
        cat2.str.endswith("-sell"),
        cat2.str.endswith("-rent"),
    ],
    [
        "temporary_rent",
        "sell",
        "rent",
    ],
    default="other"
)

display(df_pre["transaction_type"].value_counts(dropna=False))
display(pd.crosstab(df_pre["cat2_slug"], df_pre["transaction_type"]))

,count
transaction_type,
sell,597552
rent,353092
temporary_rent,29897
other,19402


transaction_type,other,rent,sell,temporary_rent
cat2_slug,,,,
commercial-rent,0,76564,0,0
commercial-sell,0,0,38860,0
real-estate-services,19402,0,0,0
residential-rent,0,276528,0,0
residential-sell,0,0,558692,0
temporary-rent,0,0,0,29897


In [ ]:
def positive_or_nan(series):
    return series.where(series > 0, np.nan)

df_pre["price_value_pos"] = positive_or_nan(df_pre["price_value"])
df_pre["rent_value_pos"] = positive_or_nan(df_pre["rent_value"])
df_pre["credit_value_pos"] = positive_or_nan(df_pre["credit_value"])
df_pre["building_size_pos"] = positive_or_nan(df_pre["building_size"])
df_pre["land_size_pos"] = positive_or_nan(df_pre["land_size"])

positive_cols = [
    "price_value_pos",
    "rent_value_pos",
    "credit_value_pos",
    "building_size_pos",
    "land_size_pos",
]

display(
    df_pre[positive_cols]
    .describe(percentiles=[0.01, 0.05, 0.5, 0.95, 0.99])
    .T
)

,count,mean,std,min,1%,5%,50%,95%,99%,max
price_value_pos,566428.0,1.742395e+10,5.888675e+11,1.0,111111.0,300000000.0,2.850000e+09,2.100000e+10,8.517300e+10,1.000000e+14
rent_value_pos,292052.0,4.934833e+10,4.176009e+12,1.0,100000.0,100000.0,6.200000e+06,6.500000e+07,2.500000e+08,1.000000e+15
credit_value_pos,349199.0,4.912485e+10,4.359308e+12,1.0,1111111.0,20000000.0,2.500000e+08,1.700000e+09,5.000000e+09,1.000000e+15
building_size_pos,999943.0,4.355842e+03,1.353702e+05,1.0,12.0,38.0,1.030000e+02,8.000000e+02,1.000000e+04,1.000000e+07
land_size_pos,186390.0,4.165602e+03,1.218947e+05,1.0,40.0,60.0,1.950000e+02,1.000000e+03,4.321430e+03,1.000000e+07


In [ ]:
df_pre["area_for_unit_price"] = df_pre["building_size_pos"]

is_plot_old = df_pre["cat3_slug"].astype("string").eq("plot-old")

df_pre.loc[
    is_plot_old & df_pre["land_size_pos"].notna(),
    "area_for_unit_price"
] = df_pre.loc[
    is_plot_old & df_pre["land_size_pos"].notna(),
    "land_size_pos"
]

display(
    df_pre[["building_size", "land_size", "area_for_unit_price"]]
    .describe(percentiles=[0.01, 0.05, 0.5, 0.95, 0.99])
    .T
)

display(
    df_pre.groupby("cat3_slug", observed=True)["area_for_unit_price"]
    .agg(count="count", mean="mean", median="median", min="min", max="max")
    .sort_values("count", ascending=False)
)

,count,mean,std,min,1%,5%,50%,95%,99%,max
building_size,999943.0,4355.841994,135370.213004,1.0,12.0,38.0,103.0,800.0,10000.00,10000000.0
land_size,186390.0,4165.601529,121894.651525,1.0,40.0,60.0,195.0,1000.0,4321.43,10000000.0
area_for_unit_price,999943.0,4355.842064,135370.213002,1.0,12.0,38.0,103.0,800.0,10000.00,10000000.0


,count,mean,median,min,max
cat3_slug,,,,,
apartment-sell,303372,499.048409,90.0,1.0,10000000.0
apartment-rent,211853,781.088325,95.0,1.0,10000000.0
plot-old,133570,10930.267066,270.0,1.0,10000000.0
house-villa-sell,121750,2543.077577,130.0,1.0,10000000.0
house-villa-rent,64675,4674.683634,100.0,1.0,10000000.0
shop-rent,45993,2233.101907,40.0,1.0,10000000.0
shop-sell,21855,1921.343674,30.0,1.0,9999996.0
office-rent,21416,2037.059862,90.0,1.0,10000000.0
suite-apartment,16460,3019.746962,80.0,1.0,8888888.0


In [ ]:
CREDIT_TO_MONTHLY_RENT_RATE = 30_000 / 1_000_000

is_rent = df_pre["transaction_type"].eq("rent")

has_rent_or_credit = (
    df_pre["rent_value_pos"].notna() |
    df_pre["credit_value_pos"].notna()
)

df_pre["monthly_rent_equivalent"] = np.where(
    is_rent & has_rent_or_credit,
    df_pre["rent_value_pos"].fillna(0) +
    df_pre["credit_value_pos"].fillna(0) * CREDIT_TO_MONTHLY_RENT_RATE,
    np.nan
)

df_pre["credit_equivalent"] = np.where(
    is_rent & has_rent_or_credit,
    df_pre["credit_value_pos"].fillna(0) +
    df_pre["rent_value_pos"].fillna(0) / CREDIT_TO_MONTHLY_RENT_RATE,
    np.nan
)

display(
    df_pre[[
        "rent_value",
        "credit_value",
        "monthly_rent_equivalent",
        "credit_equivalent"
    ]]
    .describe(percentiles=[0.01, 0.05, 0.5, 0.95, 0.99])
    .T
)

,count,mean,std,min,1%,5%,50%,95%,99%,max
rent_value,351290.0,4.102673e+10,3.807708e+12,0.000000,0.000000e+00,0.0,5.000000e+06,5.500000e+07,2.200000e+08,1.000000e+15
credit_value,352063.0,4.872522e+10,4.341543e+12,0.000000,1.000000e+06,15000000.0,2.500000e+08,1.650000e+09,5.000000e+09,1.000000e+15
monthly_rent_equivalent,352160.0,4.238673e+10,3.891019e+12,1.030000,3.321031e+05,3300000.0,1.450000e+07,1.001800e+08,3.600000e+08,1.030000e+15
credit_equivalent,352160.0,1.412891e+12,1.297006e+14,34.333333,1.107010e+07,110000000.0,4.833333e+08,3.339333e+09,1.200000e+10,3.433333e+16


In [ ]:
is_sell = df_pre["transaction_type"].eq("sell")
valid_area = df_pre["area_for_unit_price"].notna() & (df_pre["area_for_unit_price"] > 0)

df_pre["sale_price_per_m2"] = np.where(
    is_sell & df_pre["price_value_pos"].notna() & valid_area,
    df_pre["price_value_pos"] / df_pre["area_for_unit_price"],
    np.nan
)

df_pre["monthly_rent_equivalent_per_m2"] = np.where(
    is_rent & df_pre["monthly_rent_equivalent"].notna() & valid_area,
    df_pre["monthly_rent_equivalent"] / df_pre["area_for_unit_price"],
    np.nan
)

df_pre["credit_equivalent_per_m2"] = np.where(
    is_rent & df_pre["credit_equivalent"].notna() & valid_area,
    df_pre["credit_equivalent"] / df_pre["area_for_unit_price"],
    np.nan
)

# یک target کلی برای استفاده‌های بعدی
df_pre["target_price"] = np.select(
    [
        is_sell,
        is_rent,
    ],
    [
        df_pre["price_value_pos"],
        df_pre["monthly_rent_equivalent"],
    ],
    default=np.nan
)

df_pre["target_price_type"] = np.select(
    [
        is_sell,
        is_rent,
    ],
    [
        "sale_total_price",
        "monthly_rent_equivalent",
    ],
    default="not_defined"
)

df_pre["target_price_per_m2"] = np.where(
    df_pre["target_price"].notna() & valid_area,
    df_pre["target_price"] / df_pre["area_for_unit_price"],
    np.nan
)

unit_price_cols = [
    "sale_price_per_m2",
    "monthly_rent_equivalent_per_m2",
    "credit_equivalent_per_m2",
    "target_price",
    "target_price_per_m2",
]

display(
    df_pre[unit_price_cols]
    .describe(percentiles=[0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99])
    .T
)

display(df_pre["target_price_type"].value_counts(dropna=False))

,count,mean,std,min,1%,5%,25%,50%,75%,95%,99%,max
sale_price_per_m2,565174.0,1.422717e+08,7.215397e+09,0.000002,444.444000,6.000000e+05,1.280000e+07,2.780000e+07,5.060241e+07,1.500000e+08,4.000000e+08,3.333333e+12
monthly_rent_equivalent_per_m2,352160.0,1.049338e+09,1.261004e+11,0.010300,1000.000000,3.882353e+04,1.000000e+05,1.650000e+05,3.045455e+05,8.181818e+05,2.300000e+06,4.291667e+13
credit_equivalent_per_m2,352160.0,3.497793e+10,4.203346e+12,0.343333,33333.333333,1.294118e+06,3.333333e+06,5.500000e+06,1.015152e+07,2.727273e+07,7.666667e+07,1.430556e+15
target_price,917334.0,2.702193e+10,2.454880e+12,1.000000,144444.330000,5.000000e+06,1.900000e+07,1.180000e+09,3.600000e+09,1.470000e+10,5.550000e+10,1.030000e+15
target_price_per_m2,917334.0,4.904900e+08,7.833704e+10,0.000002,723.522530,6.000000e+04,2.045455e+05,8.000000e+06,3.400000e+07,1.161290e+08,3.000000e+08,4.291667e+13


,count
target_price_type,
sale_total_price,597552
monthly_rent_equivalent,353092
not_defined,49299


In [ ]:
df_pre["is_valid_geo_basic"] = (
    df_pre["location_latitude"].between(24, 40) &
    df_pre["location_longitude"].between(43, 64)
)

df_pre["is_valid_area_basic"] = (
    df_pre["area_for_unit_price"].notna() &
    df_pre["area_for_unit_price"].between(10, 20_000)
)

df_pre["is_valid_sale_price_basic"] = (
    df_pre["price_value"].isna() |
    df_pre["price_value"].between(10_000_000, 500_000_000_000)
)

df_pre["is_valid_rent_value_basic"] = (
    df_pre["rent_value"].isna() |
    df_pre["rent_value"].between(0, 1_000_000_000)
)

df_pre["is_valid_credit_value_basic"] = (
    df_pre["credit_value"].isna() |
    df_pre["credit_value"].between(0, 50_000_000_000)
)

df_pre["is_valid_target_price_basic"] = (
    df_pre["target_price"].isna() |
    (
        is_sell & df_pre["target_price"].between(10_000_000, 500_000_000_000)
    ) |
    (
        is_rent & df_pre["target_price"].between(0, 1_500_000_000)
    )
)

basic_flags = [
    "is_valid_geo_basic",
    "is_valid_area_basic",
    "is_valid_sale_price_basic",
    "is_valid_rent_value_basic",
    "is_valid_credit_value_basic",
    "is_valid_target_price_basic",
]

for col in basic_flags:
    print("=" * 80)
    print(col)
    display(df_pre[col].value_counts(dropna=False))

is_valid_geo_basic


,count
is_valid_geo_basic,
True,655576
False,344367


is_valid_area_basic


,count
is_valid_area_basic,
True,988481
False,11462


is_valid_sale_price_basic


,count
is_valid_sale_price_basic,
True,986108
False,13835


is_valid_rent_value_basic


,count
is_valid_rent_value_basic,
True,998904
False,1039


is_valid_credit_value_basic


,count
is_valid_credit_value_basic,
True,999473
False,470


is_valid_target_price_basic


,count
is_valid_target_price_basic,
True,986996
False,12947


In [ ]:
def add_group_quantile_flag(
    data,
    value_col,
    group_col,
    flag_col,
    lower_q=0.01,
    upper_q=0.99,
    min_group_count=100
):
    valid = data[value_col].notna()

    group_counts = data.loc[valid].groupby(group_col, observed=True)[value_col].transform("count")
    lower_bounds = data.loc[valid].groupby(group_col, observed=True)[value_col].transform(lambda s: s.quantile(lower_q))
    upper_bounds = data.loc[valid].groupby(group_col, observed=True)[value_col].transform(lambda s: s.quantile(upper_q))

    result = pd.Series(pd.NA, index=data.index, dtype="boolean")

    valid_index = data.loc[valid].index

    result.loc[valid_index] = (
        (group_counts >= min_group_count) &
        (data.loc[valid, value_col] >= lower_bounds) &
        (data.loc[valid, value_col] <= upper_bounds)
    ).astype("boolean")

    data[flag_col] = result
    return data

df_pre = add_group_quantile_flag(
    data=df_pre,
    value_col="area_for_unit_price",
    group_col="cat3_slug",
    flag_col="is_valid_area_q01_q99_by_cat3",
    lower_q=0.01,
    upper_q=0.99,
    min_group_count=100
)

df_pre = add_group_quantile_flag(
    data=df_pre,
    value_col="target_price_per_m2",
    group_col="cat3_slug",
    flag_col="is_valid_target_price_per_m2_q01_q99_by_cat3",
    lower_q=0.01,
    upper_q=0.99,
    min_group_count=100
)

quantile_flags = [
    "is_valid_area_q01_q99_by_cat3",
    "is_valid_target_price_per_m2_q01_q99_by_cat3",
]

for col in quantile_flags:
    print("=" * 80)
    print(col)
    display(df_pre[col].value_counts(dropna=False))

is_valid_area_q01_q99_by_cat3


,count
is_valid_area_q01_q99_by_cat3,
True,982043
False,17900


is_valid_target_price_per_m2_q01_q99_by_cat3


,count
is_valid_target_price_per_m2_q01_q99_by_cat3,
True,899050
<NA>,82609
False,18284


In [ ]:
df_pre["is_valid_for_price_analysis"] = (
    df_pre["transaction_type"].isin(["sell", "rent"]) &
    df_pre["is_valid_area_basic"] &
    df_pre["is_valid_target_price_basic"] &
    df_pre["is_valid_area_q01_q99_by_cat3"].fillna(False) &
    df_pre["is_valid_target_price_per_m2_q01_q99_by_cat3"].fillna(False)
)

display(df_pre["is_valid_for_price_analysis"].value_counts(dropna=False))

display(
    pd.crosstab(
        df_pre["cat3_slug"],
        df_pre["is_valid_for_price_analysis"],
        margins=True
    )
)

,count
is_valid_for_price_analysis,
True,877587
False,122356


is_valid_for_price_analysis,False,True,All
cat3_slug,,,
apartment-rent,7465,204388,211853
apartment-sell,12389,290983,303372
house-villa-rent,1886,62789,64675
house-villa-sell,8242,113508,121750
industry-agriculture-business-rent,1066,8089,9155
industry-agriculture-business-sell,4578,7272,11850
office-rent,1070,20346,21416
office-sell,925,4230,5155
partnership,3622,0,3622


In [ ]:
created_cols = [
    "transaction_type",
    "price_value_pos",
    "rent_value_pos",
    "credit_value_pos",
    "building_size_pos",
    "land_size_pos",
    "area_for_unit_price",
    "monthly_rent_equivalent",
    "credit_equivalent",
    "sale_price_per_m2",
    "monthly_rent_equivalent_per_m2",
    "credit_equivalent_per_m2",
    "target_price",
    "target_price_type",
    "target_price_per_m2",
    "is_valid_geo_basic",
    "is_valid_area_basic",
    "is_valid_sale_price_basic",
    "is_valid_rent_value_basic",
    "is_valid_credit_value_basic",
    "is_valid_target_price_basic",
    "is_valid_area_q01_q99_by_cat3",
    "is_valid_target_price_per_m2_q01_q99_by_cat3",
    "is_valid_for_price_analysis",
]

summary = pd.DataFrame({
    "column": created_cols,
    "dtype": [str(df_pre[col].dtype) for col in created_cols],
    "non_null": [df_pre[col].notna().sum() for col in created_cols],
    "null": [df_pre[col].isna().sum() for col in created_cols],
    "null_percent": [round(df_pre[col].isna().mean() * 100, 2) for col in created_cols],
    "unique": [df_pre[col].nunique(dropna=True) for col in created_cols],
})

display(summary)

display(
    df_pre[[
        "cat2_slug",
        "cat3_slug",
        "transaction_type",
        "price_value",
        "rent_value",
        "credit_value",
        "area_for_unit_price",
        "target_price",
        "target_price_type",
        "target_price_per_m2",
        "is_valid_for_price_analysis"
    ]].sample(20, random_state=42)
)

,column,dtype,non_null,null,null_percent,unique
0,transaction_type,object,999943,0,0.00,4
1,price_value_pos,float64,566428,433515,43.35,12942
2,rent_value_pos,float64,292052,707891,70.79,1310
3,credit_value_pos,float64,349199,650744,65.08,1420
4,building_size_pos,float64,999943,0,0.00,3572
5,land_size_pos,float64,186390,813553,81.36,1659
6,area_for_unit_price,float64,999943,0,0.00,3572
7,monthly_rent_equivalent,float64,352160,647783,64.78,5251
8,credit_equivalent,float64,352160,647783,64.78,5509
9,sale_price_per_m2,float64,565174,434769,43.48,66930


,cat2_slug,cat3_slug,transaction_type,price_value,rent_value,credit_value,area_for_unit_price,target_price,target_price_type,target_price_per_m2,is_valid_for_price_analysis
981063,residential-rent,apartment-rent,rent,NaN,2500000.0,5.000000e+07,32.0,4.000000e+06,monthly_rent_equivalent,1.250000e+05,False
305454,residential-sell,apartment-sell,sell,6.700000e+09,NaN,NaN,75.0,6.700000e+09,sale_total_price,8.933333e+07,True
232393,residential-sell,apartment-sell,sell,2.600000e+09,NaN,NaN,49.0,2.600000e+09,sale_total_price,5.306122e+07,True
36966,residential-sell,apartment-sell,sell,2.600000e+09,NaN,NaN,122.0,2.600000e+09,sale_total_price,2.131148e+07,True
944355,residential-sell,apartment-sell,sell,2.550000e+09,NaN,NaN,92.0,2.550000e+09,sale_total_price,2.771739e+07,True
382548,residential-sell,apartment-sell,sell,5.600000e+09,NaN,NaN,105.0,5.600000e+09,sale_total_price,5.333333e+07,True
63600,residential-sell,apartment-sell,sell,5.900000e+09,NaN,NaN,125.0,5.900000e+09,sale_total_price,4.720000e+07,True
489376,residential-sell,apartment-sell,sell,2.450000e+09,NaN,NaN,70.0,2.450000e+09,sale_total_price,3.500000e+07,True
190199,commercial-rent,shop-rent,rent,NaN,25000000.0,1.000000e+08,300.0,2.800000e+07,monthly_rent_equivalent,9.333333e+04,True
700155,residential-sell,apartment-sell,sell,1.880000e+09,NaN,NaN,113.0,1.880000e+09,sale_total_price,1.663717e+07,True


In [ ]:
# -----------------------------
# Refine geo validity flags
# -----------------------------

has_geo = (
    df_pre["location_latitude"].notna() &
    df_pre["location_longitude"].notna()
)

df_pre["has_geo"] = has_geo

df_pre["is_valid_geo_basic"] = pd.Series(pd.NA, index=df_pre.index, dtype="boolean")

df_pre.loc[has_geo, "is_valid_geo_basic"] = (
    df_pre.loc[has_geo, "location_latitude"].between(24, 40) &
    df_pre.loc[has_geo, "location_longitude"].between(43, 64)
)

display(df_pre["has_geo"].value_counts(dropna=False))
display(df_pre["is_valid_geo_basic"].value_counts(dropna=False))

,count
has_geo,
True,655594
False,344349


,count
is_valid_geo_basic,
True,655576
<NA>,344349
False,18


In [ ]:
preprocessing_report = pd.DataFrame({
    "item": [
        "total_rows",
        "sell_ads",
        "rent_ads",
        "temporary_rent_ads",
        "other_ads",
        "rows_with_target_price",
        "valid_for_price_analysis",
        "invalid_for_price_analysis",
        "rows_with_geo",
        "invalid_geo_rows",
        "invalid_area_basic",
        "invalid_target_price_basic"
    ],
    "count": [
        len(df_pre),
        (df_pre["transaction_type"] == "sell").sum(),
        (df_pre["transaction_type"] == "rent").sum(),
        (df_pre["transaction_type"] == "temporary_rent").sum(),
        (df_pre["transaction_type"] == "other").sum(),
        df_pre["target_price"].notna().sum(),
        df_pre["is_valid_for_price_analysis"].sum(),
        (~df_pre["is_valid_for_price_analysis"]).sum(),
        df_pre["has_geo"].sum(),
        (df_pre["is_valid_geo_basic"] == False).sum(),
        (~df_pre["is_valid_area_basic"]).sum(),
        (~df_pre["is_valid_target_price_basic"]).sum()
    ]
})

preprocessing_report["percent"] = (
    preprocessing_report["count"] / len(df_pre) * 100
).round(2)

display(preprocessing_report)

,item,count,percent
0,total_rows,999943,100.00
1,sell_ads,597552,59.76
2,rent_ads,353092,35.31
3,temporary_rent_ads,29897,2.99
4,other_ads,19402,1.94
5,rows_with_target_price,917334,91.74
6,valid_for_price_analysis,877587,87.76
7,invalid_for_price_analysis,122356,12.24
8,rows_with_geo,655594,65.56
9,invalid_geo_rows,18,0.00


In [ ]:
OUTPUT_PATH = "/content/drive/MyDrive/divar_project/cleaned_step2_benyamin.parquet"

df_pre.to_parquet(OUTPUT_PATH, index=False)

print("Saved to:", OUTPUT_PATH)
print("Shape:", df_pre.shape)
print("Columns:", df_pre.shape[1])

Saved to: /content/drive/MyDrive/divar_project/cleaned_step2_benyamin.parquet
Shape: (999943, 78)
Columns: 78


In [ ]:
df_check = pd.read_parquet(OUTPUT_PATH)

print("Loaded saved file")
print("Shape:", df_check.shape)
print("Columns:", df_check.shape[1])

new_columns_to_check = [
    "transaction_type",
    "area_for_unit_price",
    "monthly_rent_equivalent",
    "credit_equivalent",
    "sale_price_per_m2",
    "monthly_rent_equivalent_per_m2",
    "credit_equivalent_per_m2",
    "target_price",
    "target_price_type",
    "target_price_per_m2",
    "has_geo",
    "is_valid_geo_basic",
    "is_valid_for_price_analysis"
]

missing_cols = [col for col in new_columns_to_check if col not in df_check.columns]

print("Missing new columns:", missing_cols)

display(df_check[new_columns_to_check].head())

Loaded saved file
Shape: (999943, 78)
Columns: 78
Missing new columns: []


,transaction_type,area_for_unit_price,monthly_rent_equivalent,credit_equivalent,sale_price_per_m2,monthly_rent_equivalent_per_m2,credit_equivalent_per_m2,target_price,target_price_type,target_price_per_m2,has_geo,is_valid_geo_basic,is_valid_for_price_analysis
0,temporary_rent,500.0,NaN,NaN,NaN,NaN,NaN,NaN,not_defined,NaN,True,True,False
1,sell,60.0,NaN,NaN,1.416667e+08,NaN,NaN,8.500000e+09,sale_total_price,1.416667e+08,False,<NA>,True
2,rent,132.0,48500000.0,1.616667e+09,NaN,3.674242e+05,1.224747e+07,4.850000e+07,monthly_rent_equivalent,3.674242e+05,True,True,True
3,rent,90.0,123500000.0,4.116667e+09,NaN,1.372222e+06,4.574074e+07,1.235000e+08,monthly_rent_equivalent,1.372222e+06,False,<NA>,True
4,sell,115.0,NaN,NaN,5.000000e+07,NaN,NaN,5.750000e+09,sale_total_price,5.000000e+07,False,<NA>,True


تغییرات انجام شده بعد از اینکه علی یکسری تغییرات روی کد داد
.
.
.
.
.
.
.







In [ ]:
import os
import shutil
from datetime import datetime

OLD_OUTPUT_PATH = "/content/drive/MyDrive/divar_project/cleaned_step2_benyamin.parquet"

timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
BACKUP_PATH = f"/content/drive/MyDrive/divar_project/cleaned_step2_benyamin_backup_{timestamp}.parquet"

if os.path.exists(OLD_OUTPUT_PATH):
    shutil.copy2(OLD_OUTPUT_PATH, BACKUP_PATH)
    print("Backup created:")
    print(BACKUP_PATH)
else:
    print("Old output file not found. No backup created.")

Backup created:
/content/drive/MyDrive/divar_project/cleaned_step2_benyamin_backup_20260704_182912.parquet


In [ ]:
import pandas as pd
import numpy as np
import os

NEW_INPUT_PATH = "/content/drive/MyDrive/divar_project/cleaned_step1.parquet"
OLD_OUTPUT_PATH = "/content/drive/MyDrive/divar_project/cleaned_step2_benyamin.parquet"

df_new = pd.read_parquet(NEW_INPUT_PATH)
df_old_output = pd.read_parquet(OLD_OUTPUT_PATH)

print("New Ali input shape:", df_new.shape)
print("Old Benyamin output shape:", df_old_output.shape)

print("New Ali input columns:", df_new.shape[1])
print("Old Benyamin output columns:", df_old_output.shape[1])

New Ali input shape: (999943, 53)
Old Benyamin output shape: (999943, 78)
New Ali input columns: 53
Old Benyamin output columns: 78


In [ ]:
benyamin_created_cols = [
    "transaction_type",
    "price_value_pos",
    "rent_value_pos",
    "credit_value_pos",
    "building_size_pos",
    "land_size_pos",
    "area_for_unit_price",
    "monthly_rent_equivalent",
    "credit_equivalent",
    "sale_price_per_m2",
    "monthly_rent_equivalent_per_m2",
    "credit_equivalent_per_m2",
    "target_price",
    "target_price_type",
    "target_price_per_m2",
    "is_valid_geo_basic",
    "has_geo",
    "is_valid_area_basic",
    "is_valid_sale_price_basic",
    "is_valid_rent_value_basic",
    "is_valid_credit_value_basic",
    "is_valid_target_price_basic",
    "is_valid_area_q01_q99_by_cat3",
    "is_valid_target_price_per_m2_q01_q99_by_cat3",
    "is_valid_for_price_analysis",
]

already_in_new_input = [col for col in benyamin_created_cols if col in df_new.columns]

print("Benyamin columns already existing in new Ali input:")
print(already_in_new_input)

Benyamin columns already existing in new Ali input:
[]


In [ ]:
old_base = df_old_output.drop(
    columns=[col for col in benyamin_created_cols if col in df_old_output.columns],
    errors="ignore"
)

new_cols = set(df_new.columns)
old_base_cols = set(old_base.columns)

added_cols = sorted(new_cols - old_base_cols)
removed_cols = sorted(old_base_cols - new_cols)
common_cols = sorted(new_cols & old_base_cols)

print("Columns added by Ali:")
print(added_cols)

print("\nColumns removed by Ali:")
print(removed_cols)

dtype_compare = pd.DataFrame({
    "column": common_cols,
    "old_dtype": [str(old_base[col].dtype) for col in common_cols],
    "new_dtype": [str(df_new[col].dtype) for col in common_cols],
})

dtype_changed = dtype_compare[dtype_compare["old_dtype"] != dtype_compare["new_dtype"]]

print("\nDtype changes:")
display(dtype_changed)

Columns added by Ali:
[]

Columns removed by Ali:
[]

Dtype changes:


,column,old_dtype,new_dtype


In [ ]:
required_cols_for_benyamin = [
    "cat2_slug",
    "cat3_slug",
    "price_value",
    "rent_value",
    "credit_value",
    "building_size",
    "land_size",
    "location_latitude",
    "location_longitude",
]

missing_required = [col for col in required_cols_for_benyamin if col not in df_new.columns]

if missing_required:
    print("Missing required columns:")
    print(missing_required)
else:
    print("All required columns are available.")

All required columns are available.


In [ ]:
def make_schema_summary(data, name):
    return pd.DataFrame({
        "column": data.columns,
        f"{name}_dtype": [str(data[col].dtype) for col in data.columns],
        f"{name}_null_percent": [round(data[col].isna().mean() * 100, 2) for col in data.columns],
        f"{name}_unique": [data[col].nunique(dropna=True) for col in data.columns],
    })

old_schema = make_schema_summary(old_base, "old")
new_schema = make_schema_summary(df_new, "new")

schema_diff = old_schema.merge(new_schema, on="column", how="outer")

schema_diff["null_percent_diff"] = (
    schema_diff["new_null_percent"] - schema_diff["old_null_percent"]
).round(2)

display(
    schema_diff[
        schema_diff["column"].isin(required_cols_for_benyamin)
    ].sort_values("column")
)

,column,old_dtype,old_null_percent,old_unique,new_dtype,new_null_percent,new_unique,null_percent_diff
1,building_size,float64,0.00,3572,float64,0.00,3572,0.0
2,cat2_slug,category,0.00,6,category,0.00,6,0.0
3,cat3_slug,category,0.00,16,category,0.00,16,0.0
10,credit_value,float64,64.79,1421,float64,64.79,1421,0.0
30,land_size,float64,81.36,1659,float64,81.36,1659,0.0
31,location_latitude,float64,34.44,387482,float64,34.44,387482,0.0
32,location_longitude,float64,34.44,429426,float64,34.44,429426,0.0
36,price_value,float64,43.16,12943,float64,43.16,12943,0.0
42,rent_value,float64,64.87,1311,float64,64.87,1311,0.0


In [ ]:
check_cols = [
    "cat2_slug",
    "cat3_slug",
    "price_value",
    "rent_value",
    "credit_value",
    "building_size",
    "land_size",
]

for col in check_cols:
    if col not in df_new.columns:
        continue

    print("=" * 100)
    print(col)
    print("dtype:", df_new[col].dtype)
    print("missing:", df_new[col].isna().sum(), "|", round(df_new[col].isna().mean() * 100, 2), "%")

    if pd.api.types.is_numeric_dtype(df_new[col]):
        display(
            df_new[col]
            .describe(percentiles=[0.01, 0.05, 0.5, 0.95, 0.99])
            .to_frame()
            .T
        )
    else:
        display(df_new[col].value_counts(dropna=False).head(20))

cat2_slug
dtype: category
missing: 0 | 0.0 %


,count
cat2_slug,
residential-sell,558692
residential-rent,276528
commercial-rent,76564
commercial-sell,38860
temporary-rent,29897
real-estate-services,19402


cat3_slug
dtype: category
missing: 0 | 0.0 %


,count
cat3_slug,
apartment-sell,303372
apartment-rent,211853
plot-old,133570
house-villa-sell,121750
house-villa-rent,64675
shop-rent,45993
shop-sell,21855
office-rent,21416
suite-apartment,16460


price_value
dtype: float64
missing: 431613 | 43.16 %


,count,mean,std,min,1%,5%,50%,95%,99%,max
price_value,568330.0,1.736564e+10,5.878822e+11,0.0,3400.0,260000000.0,2.840000e+09,2.100000e+10,8.500000e+10,1.000000e+14


rent_value
dtype: float64
missing: 648653 | 64.87 %


,count,mean,std,min,1%,5%,50%,95%,99%,max
rent_value,351290.0,4.102673e+10,3.807708e+12,0.0,0.0,0.0,5000000.0,55000000.0,220000000.0,1.000000e+15


credit_value
dtype: float64
missing: 647880 | 64.79 %


,count,mean,std,min,1%,5%,50%,95%,99%,max
credit_value,352063.0,4.872522e+10,4.341543e+12,0.0,1000000.0,15000000.0,250000000.0,1.650000e+09,5.000000e+09,1.000000e+15


building_size
dtype: float64
missing: 0 | 0.0 %


,count,mean,std,min,1%,5%,50%,95%,99%,max
building_size,999943.0,4356.495411,135370.192797,1.0,12.0,38.0,105.0,800.0,10000.0,10000000.0


land_size
dtype: float64
missing: 813553 | 81.36 %


,count,mean,std,min,1%,5%,50%,95%,99%,max
land_size,186390.0,4165.601529,121894.651525,1.0,40.0,60.0,195.0,1000.0,4321.43,10000000.0


## Rerun Benyamin Preprocessing on Updated Ali Data

در این بخش، preprocessing مربوط به قیمت واحد و outlierها دوباره روی نسخه‌ی به‌روزشده‌ی دیتای علی اجرا می‌شود.

از آنجا که علی مدیریت مقادیر گم‌شده را اصلاح کرده است، یعنی `construction_year` را با میانه کل و `rooms_count` و `building_size` را با میانه‌ی هر `cat3_slug` پر کرده، ستون‌هایی که به متراژ و قیمت وابسته‌اند باید دوباره محاسبه شوند.

در این نوت‌بوک، اگر دیتای جدید علی قبلاً با نام `df_new` لود شده باشد، همان به عنوان ورودی استفاده می‌شود. در غیر این صورت، باید فایل خروجی علی یعنی نسخه‌ی جدید `cleaned_step1.parquet` به صورت دستی لود شود.

در این مرحله هیچ رکوردی حذف نمی‌شود؛ فقط ستون‌های کمکی و flagهای اعتبارسنجی ساخته می‌شوند. خروجی نهایی با نام `cleaned_step2_benyamin_v2.parquet` ذخیره می‌شود.

In [3]:
import pandas as pd
import numpy as np
import os
import glob
import gc

pd.set_option("display.max_columns", 150)
pd.set_option("display.max_rows", 100)
pd.set_option("display.width", 250)

# اگر Drive هنوز mount نشده باشد، mount می‌کنیم
if not os.path.exists("/content/drive/MyDrive"):
    from google.colab import drive
    drive.mount("/content/drive", force_remount=True)

# اگر از قبل دیتافریم در حافظه وجود داشته باشد، همان را استفاده می‌کنیم
if "df_new" in globals() and isinstance(df_new, pd.DataFrame):
    df_base_v2 = df_new.copy(deep=False)
    print("Using Ali updated data from existing df_new.")

elif "df" in globals() and isinstance(df, pd.DataFrame):
    df_base_v2 = df.copy(deep=False)
    print("Using existing df as base data.")

else:
    print("No existing dataframe found in memory.")
    print("Searching for Ali cleaned parquet file in Google Drive...")

    candidate_paths = [
        "/content/drive/MyDrive/divar_project/cleaned_step1.parquet",
        "/content/drive/MyDrive/divar_project/cleaned_step1_v2.parquet",
        "/content/drive/MyDrive/divar_project/cleaned_step1_ali.parquet",
        "/content/drive/MyDrive/divar_project/cleaned_ali.parquet",
    ]

    existing_candidates = [path for path in candidate_paths if os.path.exists(path)]

    if len(existing_candidates) == 0:
        matches = glob.glob(
            "/content/drive/MyDrive/**/*.parquet",
            recursive=True
        )

        matches = [
            path for path in matches
            if (
                "clean" in os.path.basename(path).lower()
                or "step1" in os.path.basename(path).lower()
                or "ali" in os.path.basename(path).lower()
            )
        ]

        print("Possible parquet files found:")
        for i, path in enumerate(matches):
            print(i, ":", path)

        if len(matches) == 0:
            raise FileNotFoundError(
                "No cleaned parquet file was found in Google Drive. "
                "First run Ali preprocessing notebook or ask Ali for the exact output file path."
            )

        if len(matches) > 1:
            raise ValueError(
                "Multiple possible parquet files found. "
                "Choose the correct path from the printed list and set INPUT_PATH manually."
            )

        INPUT_PATH = matches[0]

    else:
        INPUT_PATH = existing_candidates[0]

    print("Loading input file:")
    print(INPUT_PATH)

    df_base_v2 = pd.read_parquet(INPUT_PATH)

print("Base data loaded.")
print("Shape:", df_base_v2.shape)
print("Columns:", df_base_v2.shape[1])

display(df_base_v2.head(3))

No existing dataframe found in memory.
Searching for Ali cleaned parquet file in Google Drive...
Loading input file:
/content/drive/MyDrive/divar_project/cleaned_step1.parquet
Base data loaded.
Shape: (999943, 53)
Columns: 53


,cat2_slug,cat3_slug,city_slug,neighborhood_slug,created_at_month,user_type,description,title,rent_mode,rent_value,rent_type,price_mode,price_value,credit_mode,credit_value,rent_credit_transform,transformable_price,transformable_credit,transformed_credit,transformable_rent,transformed_rent,land_size,building_size,deed_type,has_business_deed,floor,rooms_count,total_floors_count,unit_per_floor,has_balcony,has_elevator,has_warehouse,has_parking,construction_year,is_rebuilt,has_warm_water_provider,has_heating_system,has_cooling_system,has_restroom,has_security_guard,has_barbecue,building_direction,has_pool,has_jacuzzi,has_sauna,floor_material,property_type,regular_person_capacity,location_latitude,location_longitude,location_radius,created_at_shamsi,created_at_shamsi_readable
0,temporary-rent,villa,karaj,mehrshahr,2024-08-01,مشاور املاک,۵۰۰متر\n۲۰۰متر بنا دوبلکس\n۳خواب\nاستخر آبگرم ...,باغ ویلا اجاره روزانه استخر داخل لشکرآباد سهیلیه,NaN,NaN,NaN,NaN,NaN,NaN,NaN,<NA>,<NA>,NaN,NaN,NaN,NaN,NaN,500.0,NaN,<NA>,NaN,3.0,NaN,NaN,False,False,False,False,1395.0,False,NaN,NaN,NaN,NaN,False,False,NaN,False,False,False,NaN,NaN,4.0,35.811684,50.936600,500.0,1403-05,مرداد 1403
1,residential-sell,apartment-sell,tehran,gholhak,2024-05-01,مشاور املاک,دسترسی عالی به مترو و شریعتی \nمشاعات تمیز \nب...,۶۰ متر قلهک فول امکانات,NaN,NaN,NaN,مقطوع,8.500000e+09,NaN,NaN,<NA>,<NA>,NaN,NaN,NaN,NaN,NaN,60.0,NaN,<NA>,3.0,1.0,NaN,NaN,False,True,True,True,1384.0,False,NaN,NaN,NaN,NaN,False,False,NaN,False,False,False,NaN,NaN,NaN,NaN,NaN,500.0,1403-02,اردیبهشت 1403
2,residential-rent,apartment-rent,tehran,tohid,2024-10-01,NaN,تخلیه پایان ماه,آپارتمان ۳ خوابه ۱۳۲ متر,مقطوع,26000000.0,NaN,NaN,NaN,مقطوع,750000000.0,False,False,750000000.0,NaN,26000000.0,NaN,NaN,132.0,NaN,<NA>,3.0,3.0,NaN,NaN,False,True,True,True,1401.0,False,NaN,NaN,NaN,NaN,False,False,NaN,False,False,False,NaN,NaN,NaN,35.703865,51.373459,NaN,1403-07,مهر 1403


In [4]:
required_cols = [
    "cat2_slug",
    "cat3_slug",
    "price_value",
    "rent_value",
    "credit_value",
    "building_size",
    "land_size",
    "location_latitude",
    "location_longitude"
]

missing_required = [col for col in required_cols if col not in df_base_v2.columns]

print("Missing required columns:", missing_required)

if missing_required:
    raise ValueError("Some required columns are missing. Stop here and check the input data.")

Missing required columns: []


In [5]:
df_pre_v2 = df_base_v2.copy(deep=False)

print("Base shape:", df_base_v2.shape)
print("Preprocessing v2 shape:", df_pre_v2.shape)

Base shape: (999943, 53)
Preprocessing v2 shape: (999943, 53)


In [6]:
# -----------------------------
# 1. Transaction type
# -----------------------------

cat2 = df_pre_v2["cat2_slug"].astype("string")

df_pre_v2["transaction_type"] = np.select(
    [
        cat2.eq("temporary-rent"),
        cat2.str.endswith("-sell"),
        cat2.str.endswith("-rent"),
    ],
    [
        "temporary_rent",
        "sell",
        "rent",
    ],
    default="other"
)

is_sell = df_pre_v2["transaction_type"].eq("sell")
is_rent = df_pre_v2["transaction_type"].eq("rent")


# -----------------------------
# 2. Positive versions of numeric columns
# -----------------------------

def positive_or_nan(series):
    return series.where(series > 0, np.nan)

df_pre_v2["price_value_pos"] = positive_or_nan(df_pre_v2["price_value"])
df_pre_v2["rent_value_pos"] = positive_or_nan(df_pre_v2["rent_value"])
df_pre_v2["credit_value_pos"] = positive_or_nan(df_pre_v2["credit_value"])
df_pre_v2["building_size_pos"] = positive_or_nan(df_pre_v2["building_size"])
df_pre_v2["land_size_pos"] = positive_or_nan(df_pre_v2["land_size"])


# -----------------------------
# 3. Area used for unit price
# -----------------------------

df_pre_v2["area_for_unit_price"] = df_pre_v2["building_size_pos"]

is_plot_old = df_pre_v2["cat3_slug"].astype("string").eq("plot-old")

df_pre_v2.loc[
    is_plot_old & df_pre_v2["land_size_pos"].notna(),
    "area_for_unit_price"
] = df_pre_v2.loc[
    is_plot_old & df_pre_v2["land_size_pos"].notna(),
    "land_size_pos"
]


# -----------------------------
# 4. Rent-credit equivalent values
# -----------------------------

CREDIT_TO_MONTHLY_RENT_RATE = 30_000 / 1_000_000

has_rent_or_credit = (
    df_pre_v2["rent_value_pos"].notna() |
    df_pre_v2["credit_value_pos"].notna()
)

df_pre_v2["monthly_rent_equivalent"] = np.where(
    is_rent & has_rent_or_credit,
    df_pre_v2["rent_value_pos"].fillna(0) +
    df_pre_v2["credit_value_pos"].fillna(0) * CREDIT_TO_MONTHLY_RENT_RATE,
    np.nan
)

df_pre_v2["credit_equivalent"] = np.where(
    is_rent & has_rent_or_credit,
    df_pre_v2["credit_value_pos"].fillna(0) +
    df_pre_v2["rent_value_pos"].fillna(0) / CREDIT_TO_MONTHLY_RENT_RATE,
    np.nan
)


# -----------------------------
# 5. Unit price and target price
# -----------------------------

valid_area = (
    df_pre_v2["area_for_unit_price"].notna() &
    (df_pre_v2["area_for_unit_price"] > 0)
)

df_pre_v2["sale_price_per_m2"] = np.where(
    is_sell & df_pre_v2["price_value_pos"].notna() & valid_area,
    df_pre_v2["price_value_pos"] / df_pre_v2["area_for_unit_price"],
    np.nan
)

df_pre_v2["monthly_rent_equivalent_per_m2"] = np.where(
    is_rent & df_pre_v2["monthly_rent_equivalent"].notna() & valid_area,
    df_pre_v2["monthly_rent_equivalent"] / df_pre_v2["area_for_unit_price"],
    np.nan
)

df_pre_v2["credit_equivalent_per_m2"] = np.where(
    is_rent & df_pre_v2["credit_equivalent"].notna() & valid_area,
    df_pre_v2["credit_equivalent"] / df_pre_v2["area_for_unit_price"],
    np.nan
)

df_pre_v2["target_price"] = np.select(
    [
        is_sell,
        is_rent,
    ],
    [
        df_pre_v2["price_value_pos"],
        df_pre_v2["monthly_rent_equivalent"],
    ],
    default=np.nan
)

df_pre_v2["target_price_type"] = np.select(
    [
        is_sell,
        is_rent,
    ],
    [
        "sale_total_price",
        "monthly_rent_equivalent",
    ],
    default="not_defined"
)

df_pre_v2["target_price_per_m2"] = np.where(
    df_pre_v2["target_price"].notna() & valid_area,
    df_pre_v2["target_price"] / df_pre_v2["area_for_unit_price"],
    np.nan
)


# -----------------------------
# 6. Basic validation flags
# -----------------------------

df_pre_v2["is_valid_area_basic"] = (
    df_pre_v2["area_for_unit_price"].notna() &
    df_pre_v2["area_for_unit_price"].between(10, 20_000)
)

df_pre_v2["is_valid_sale_price_basic"] = (
    df_pre_v2["price_value"].isna() |
    df_pre_v2["price_value"].between(10_000_000, 500_000_000_000)
)

df_pre_v2["is_valid_rent_value_basic"] = (
    df_pre_v2["rent_value"].isna() |
    df_pre_v2["rent_value"].between(0, 1_000_000_000)
)

df_pre_v2["is_valid_credit_value_basic"] = (
    df_pre_v2["credit_value"].isna() |
    df_pre_v2["credit_value"].between(0, 50_000_000_000)
)

df_pre_v2["is_valid_target_price_basic"] = (
    df_pre_v2["target_price"].isna() |
    (
        is_sell &
        df_pre_v2["target_price"].between(10_000_000, 500_000_000_000)
    ) |
    (
        is_rent &
        df_pre_v2["target_price"].between(0, 1_500_000_000)
    )
)


# -----------------------------
# 7. Quantile-based validation flags by cat3_slug
# -----------------------------

def add_group_quantile_flag(
    data,
    value_col,
    group_col,
    flag_col,
    lower_q=0.01,
    upper_q=0.99,
    min_group_count=100
):
    valid = data[value_col].notna()

    group_counts = (
        data.loc[valid]
        .groupby(group_col, observed=True)[value_col]
        .transform("count")
    )

    lower_bounds = (
        data.loc[valid]
        .groupby(group_col, observed=True)[value_col]
        .transform(lambda s: s.quantile(lower_q))
    )

    upper_bounds = (
        data.loc[valid]
        .groupby(group_col, observed=True)[value_col]
        .transform(lambda s: s.quantile(upper_q))
    )

    result = pd.Series(pd.NA, index=data.index, dtype="boolean")
    valid_index = data.loc[valid].index

    result.loc[valid_index] = (
        (group_counts >= min_group_count) &
        (data.loc[valid, value_col] >= lower_bounds) &
        (data.loc[valid, value_col] <= upper_bounds)
    ).astype("boolean")

    data[flag_col] = result
    return data


df_pre_v2 = add_group_quantile_flag(
    data=df_pre_v2,
    value_col="area_for_unit_price",
    group_col="cat3_slug",
    flag_col="is_valid_area_q01_q99_by_cat3",
    lower_q=0.01,
    upper_q=0.99,
    min_group_count=100
)

df_pre_v2 = add_group_quantile_flag(
    data=df_pre_v2,
    value_col="target_price_per_m2",
    group_col="cat3_slug",
    flag_col="is_valid_target_price_per_m2_q01_q99_by_cat3",
    lower_q=0.01,
    upper_q=0.99,
    min_group_count=100
)


# -----------------------------
# 8. Final price-analysis validation flag
# -----------------------------

df_pre_v2["is_valid_for_price_analysis"] = (
    df_pre_v2["transaction_type"].isin(["sell", "rent"]) &
    df_pre_v2["is_valid_area_basic"] &
    df_pre_v2["is_valid_target_price_basic"] &
    df_pre_v2["is_valid_area_q01_q99_by_cat3"].fillna(False) &
    df_pre_v2["is_valid_target_price_per_m2_q01_q99_by_cat3"].fillna(False)
)


# -----------------------------
# 9. Geo flags
# -----------------------------

has_geo = (
    df_pre_v2["location_latitude"].notna() &
    df_pre_v2["location_longitude"].notna()
)

df_pre_v2["has_geo"] = has_geo

df_pre_v2["is_valid_geo_basic"] = pd.Series(
    pd.NA,
    index=df_pre_v2.index,
    dtype="boolean"
)

df_pre_v2.loc[has_geo, "is_valid_geo_basic"] = (
    df_pre_v2.loc[has_geo, "location_latitude"].between(24, 40) &
    df_pre_v2.loc[has_geo, "location_longitude"].between(43, 64)
)


print("Benyamin preprocessing v2 completed.")
print("Shape:", df_pre_v2.shape)
print("Columns:", df_pre_v2.shape[1])

Benyamin preprocessing v2 completed.
Shape: (999943, 78)
Columns: 78


In [7]:
preprocessing_report_v2 = pd.DataFrame({
    "item": [
        "total_rows",
        "sell_ads",
        "rent_ads",
        "temporary_rent_ads",
        "other_ads",
        "rows_with_target_price",
        "valid_for_price_analysis",
        "invalid_for_price_analysis",
        "rows_with_geo",
        "invalid_geo_rows",
        "invalid_area_basic",
        "invalid_target_price_basic"
    ],
    "count": [
        len(df_pre_v2),
        (df_pre_v2["transaction_type"] == "sell").sum(),
        (df_pre_v2["transaction_type"] == "rent").sum(),
        (df_pre_v2["transaction_type"] == "temporary_rent").sum(),
        (df_pre_v2["transaction_type"] == "other").sum(),
        df_pre_v2["target_price"].notna().sum(),
        df_pre_v2["is_valid_for_price_analysis"].sum(),
        (~df_pre_v2["is_valid_for_price_analysis"]).sum(),
        df_pre_v2["has_geo"].sum(),
        (df_pre_v2["is_valid_geo_basic"] == False).sum(),
        (~df_pre_v2["is_valid_area_basic"]).sum(),
        (~df_pre_v2["is_valid_target_price_basic"]).sum()
    ]
})

preprocessing_report_v2["percent"] = (
    preprocessing_report_v2["count"] / len(df_pre_v2) * 100
).round(2)

display(preprocessing_report_v2)

,item,count,percent
0,total_rows,999943,100.00
1,sell_ads,597552,59.76
2,rent_ads,353092,35.31
3,temporary_rent_ads,29897,2.99
4,other_ads,19402,1.94
5,rows_with_target_price,917334,91.74
6,valid_for_price_analysis,877587,87.76
7,invalid_for_price_analysis,122356,12.24
8,rows_with_geo,655594,65.56
9,invalid_geo_rows,18,0.00


In [8]:
created_cols_v2 = [
    "transaction_type",
    "price_value_pos",
    "rent_value_pos",
    "credit_value_pos",
    "building_size_pos",
    "land_size_pos",
    "area_for_unit_price",
    "monthly_rent_equivalent",
    "credit_equivalent",
    "sale_price_per_m2",
    "monthly_rent_equivalent_per_m2",
    "credit_equivalent_per_m2",
    "target_price",
    "target_price_type",
    "target_price_per_m2",
    "has_geo",
    "is_valid_geo_basic",
    "is_valid_area_basic",
    "is_valid_sale_price_basic",
    "is_valid_rent_value_basic",
    "is_valid_credit_value_basic",
    "is_valid_target_price_basic",
    "is_valid_area_q01_q99_by_cat3",
    "is_valid_target_price_per_m2_q01_q99_by_cat3",
    "is_valid_for_price_analysis",
]

missing_created_cols = [
    col for col in created_cols_v2
    if col not in df_pre_v2.columns
]

print("Missing created columns:", missing_created_cols)

summary_v2 = pd.DataFrame({
    "column": created_cols_v2,
    "dtype": [str(df_pre_v2[col].dtype) for col in created_cols_v2],
    "non_null": [df_pre_v2[col].notna().sum() for col in created_cols_v2],
    "null": [df_pre_v2[col].isna().sum() for col in created_cols_v2],
    "null_percent": [round(df_pre_v2[col].isna().mean() * 100, 2) for col in created_cols_v2],
    "unique": [df_pre_v2[col].nunique(dropna=True) for col in created_cols_v2],
})

display(summary_v2)

Missing created columns: []


,column,dtype,non_null,null,null_percent,unique
0,transaction_type,object,999943,0,0.00,4
1,price_value_pos,float64,566428,433515,43.35,12942
2,rent_value_pos,float64,292052,707891,70.79,1310
3,credit_value_pos,float64,349199,650744,65.08,1420
4,building_size_pos,float64,999943,0,0.00,3572
5,land_size_pos,float64,186390,813553,81.36,1659
6,area_for_unit_price,float64,999943,0,0.00,3572
7,monthly_rent_equivalent,float64,352160,647783,64.78,5251
8,credit_equivalent,float64,352160,647783,64.78,5509
9,sale_price_per_m2,float64,565174,434769,43.48,66925


In [9]:
OUTPUT_PATH_V2 = "/content/drive/MyDrive/divar_project/cleaned_step2_benyamin_v2.parquet"

output_dir = os.path.dirname(OUTPUT_PATH_V2)

if output_dir and not os.path.exists(output_dir):
    os.makedirs(output_dir, exist_ok=True)

df_pre_v2.to_parquet(OUTPUT_PATH_V2, index=False)

print("Saved to:", OUTPUT_PATH_V2)
print("Shape:", df_pre_v2.shape)
print("Columns:", df_pre_v2.shape[1])

Saved to: /content/drive/MyDrive/divar_project/cleaned_step2_benyamin_v2.parquet
Shape: (999943, 78)
Columns: 78


In [10]:
df_check_v2 = pd.read_parquet(OUTPUT_PATH_V2)

print("Loaded saved v2 file")
print("Shape:", df_check_v2.shape)
print("Columns:", df_check_v2.shape[1])

new_columns_to_check = [
    "transaction_type",
    "area_for_unit_price",
    "monthly_rent_equivalent",
    "credit_equivalent",
    "sale_price_per_m2",
    "monthly_rent_equivalent_per_m2",
    "credit_equivalent_per_m2",
    "target_price",
    "target_price_type",
    "target_price_per_m2",
    "has_geo",
    "is_valid_geo_basic",
    "is_valid_for_price_analysis"
]

missing_cols = [
    col for col in new_columns_to_check
    if col not in df_check_v2.columns
]

print("Missing new columns:", missing_cols)

display(df_check_v2[new_columns_to_check].head())

Loaded saved v2 file
Shape: (999943, 78)
Columns: 78
Missing new columns: []


,transaction_type,area_for_unit_price,monthly_rent_equivalent,credit_equivalent,sale_price_per_m2,monthly_rent_equivalent_per_m2,credit_equivalent_per_m2,target_price,target_price_type,target_price_per_m2,has_geo,is_valid_geo_basic,is_valid_for_price_analysis
0,temporary_rent,500.0,NaN,NaN,NaN,NaN,NaN,NaN,not_defined,NaN,True,True,False
1,sell,60.0,NaN,NaN,1.416667e+08,NaN,NaN,8.500000e+09,sale_total_price,1.416667e+08,False,<NA>,True
2,rent,132.0,48500000.0,1.616667e+09,NaN,3.674242e+05,1.224747e+07,4.850000e+07,monthly_rent_equivalent,3.674242e+05,True,True,True
3,rent,90.0,123500000.0,4.116667e+09,NaN,1.372222e+06,4.574074e+07,1.235000e+08,monthly_rent_equivalent,1.372222e+06,False,<NA>,True
4,sell,115.0,NaN,NaN,5.000000e+07,NaN,NaN,5.750000e+09,sale_total_price,5.000000e+07,False,<NA>,True


### نتیجه اجرای مجدد Preprocessing

پس از به‌روزرسانی دیتای علی، preprocessing مربوط به قیمت واحد و outlierها دوباره اجرا شد.

در نسخه‌ی جدید، مقادیر گم‌شده‌ی `construction_year` با میانه کل و مقادیر گم‌شده‌ی `rooms_count` و `building_size` با میانه‌ی هر `cat3_slug` پر شده‌اند. به همین دلیل، ستون‌های وابسته به متراژ و قیمت دوباره از روی دیتای جدید محاسبه شدند.

در این مرحله هیچ رکوردی حذف نشد. فقط ستون‌های کمکی و flagهای اعتبارسنجی ساخته شدند. خروجی نهایی با نام `cleaned_step2_benyamin_v2.parquet` ذخیره شد.

خروجی نهایی شامل 999,943 رکورد و 78 ستون است. از این تعداد، 917,334 رکورد دارای `target_price` هستند و 877,587 رکورد، یعنی حدود 87.76٪ کل داده، برای تحلیل قیمت معتبر تشخیص داده شدند.